sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/NexmarkLauncher.java

```diff
@@ -760,66 +763,69 @@ private String logsDir(long now) {
+    eventToBytes.apply(
+        KafkaIO.<Void, byte[]>write()
+            .withBootstrapServers(options.getBootstrapServers())
+            .withTopic(options.getKafkaEventsTopic())
+            .withValueSerializer(ByteArraySerializer.class)
+            .values());
   }
 
-
   static final DoFn<KV<Long, byte[]>, Event> BYTEARRAY_TO_EVENT =
-          new DoFn<KV<Long, byte[]>, Event>() {
-            @ProcessElement
-            public void processElement(ProcessContext c) {
-              byte[] encodedEvent = c.element().getValue();
-              try {
-                Event event = CoderUtils.decodeFromByteArray(Event.CODER, encodedEvent);
-                c.output(event);
-              } catch (CoderException e) {
-                LOG.error("Error while decoding Event from Kafka message: serialization error");
-              }
-            }
-          };
+      new DoFn<KV<Long, byte[]>, Event>() {
+        @ProcessElement
+        public void processElement(ProcessContext c) {
+          byte[] encodedEvent = c.element().getValue();
+          try {
+            Event event = CoderUtils.decodeFromByteArray(Event.CODER, encodedEvent);
+            c.output(event);
+          } catch (CoderException e) {
+            LOG.error("Error while decoding Event from Kafka message: serialization error");
+          }
+        }
+      };
 
   /**
    * Return source of events from Kafka.
    */
-  private PCollection<Event> sourceEventsFromKafka(Pipeline p) {
-    NexmarkUtils.console("Reading events from Kafka Topic %s", options.getKafkaSourceTopic());
-
-    checkArgument(!Strings.isNullOrEmpty(options.getBootstrapServers()),
-        "Missing --bootstrapServers");
+  private PCollection<Event> sourceEventsFromKafka(Pipeline p, final Instant now) {
+    NexmarkUtils.console("Reading events from Kafka Topic %s", options.getKafkaEventsTopic());
+    checkArgument((options.getBootstrapServers() != null), "Missing --bootstrapServers");
 
-    KafkaIO.Read<Long, byte[]> read = KafkaIO.<Long, byte[]>read()
+    KafkaIO.Read<Long, byte[]> read =
+        KafkaIO.<Long, byte[]>read()
             .withBootstrapServers(options.getBootstrapServers())
-            .withTopic(options.getKafkaSourceTopic())
+            .withTopic(options.getKafkaEventsTopic())
             .withKeyDeserializer(LongDeserializer.class)
-            .withValueDeserializer(ByteArrayDeserializer.class);
+            .withValueDeserializer(ByteArrayDeserializer.class)
+            .withStartReadTime(now)
+            .withMaxNumRecords(
+                options.getNumEvents() != null ? options.getNumEvents() : Long.MAX_VALUE);
 
-    return p
-      .apply(queryName + ".ReadKafkaEvents", read.withoutMetadata())
-      .apply(queryName + ".KafkaToEvents", ParDo.of(BYTEARRAY_TO_EVENT));
+    return p.apply(queryName + ".ReadKafkaEvents", read.withoutMetadata())
+            .apply(queryName + ".KafkaToEvents", ParDo.of(BYTEARRAY_TO_EVENT));
```

Comment:
I think earlier indentation is more consistent with rest of Beam ('.' with 4 spaces).

Reply:
Done